### PR TITLE
[Proposal] Add Platforms editing to Project Configuration

### PR DESCRIPTION
**Pull request by serjooo**
### Description
This PR proposes to add support to edit all `localComponents` default supported platforms. The use case for this arises with the deprecation of a platform version where you want to update all packages supported platforms. For example, we deprecated `iOS .v14` recently in our project, and we would like to update all packages to `iOS .v15`. You can easily do this in Xcode doing a find and replace; however, it is not as trivial while using Phoenix as you would need to go and edit each Component manually. To make Phoenix better, some trivial tasks as this one should be as easy as doing it on Xcode.

Would love to hear thoughts and feedback, and maybe any build up on this idea.

### Changes
There are a couple of major changes done in the project that may or may not be necessary:

1. We extracted the platforms from being individual properties into a struct called `Platforms`. This was done to expose the same properties into `ProjectConfiguration` without much duplication.
2. We extracted the platforms selection menu to its own view called `PlatformsEditingView`. This was done to reuse the same menus in the `ProjectConfigiration` edit view and `Component` edit view
3. Decoding and Encoding. With `Platforms` being its own struct we made the Decoding and Encoding of `Component` backwards compatible to adapt these. However, Project Configuration uses the plain `Platforms` nesting and adds the new properties

### Resources

I already have this PR working and edits to all local components are working. A preview of the change looks like this:

Screenshot 2023-09-23 at 18 16 13

### CODE REVIEW

File: Modules/Entities/PhoenixDocument/Sources/PhoenixDocument/Component/ProjectConfiguration.swift [hunk 1]
```diff
@@ -39,6 +43,8 @@ public struct ProjectConfiguration: Codable, Hashable {
         macrosFolderName = try container.decodeIfPresent(String.self, forKey: .macrosFolderName) ?? ProjectConfigurationConstants.defaultMacrosFolderName
         swiftVersion = (try? container.decode(String.self, forKey: .swiftVersion)) ?? "5.9"
         defaultOrganizationIdentifier = try? container.decodeIfPresent(String.self, forKey: .defaultOrganizationIdentifier)
+        platforms = (try? container.decode(Component.Platforms.self, forKey: .platforms)) ?? .empty
+        print(platforms)
```
Comment (Tawa) [inline Modules/Entities/PhoenixDocument/Sources/PhoenixDocument/Component/ProjectConfiguration.swift:47]: You forgot this print statement here.

File: Phoenix/PhoenixDocument+Extensions.swift [hunk 2]
```diff
@@ -17,7 +17,9 @@ extension PhoenixDocument: FileDocument {
     public init(configuration: ReadConfiguration) throws {
         if configuration.file.isDirectory, let fileWrapper = configuration.file.fileWrappers {
             let phoenixDocumentFileWrappersDecoder = Container.phoenixDocumentFileWrappersDecoder()
-            self = try phoenixDocumentFileWrappersDecoder.phoenixDocument(from: fileWrapper)
+            let phoenixDocument = try phoenixDocumentFileWrappersDecoder.phoenixDocument(from: fileWrapper)
+            self = phoenixDocument
+            print(self)
```
Comment (Tawa) [inline Phoenix/PhoenixDocument+Extensions.swift:22]: I'm guessing this change is not necessary?

**Review by Tawa — COMMENTED** [review]
Hey Serjooo,

Thank you for the work you put into this feature. I think it's an excellent idea.
I do however think the User Experience should be a bit different.

If I understood correctly from the code, your approach makes the change you make in the project configuration apply to all the components in the project. So when you update the iOS Version in Project Configuration, all components will have this new iOS Version set to them.

I would rather take have this behave similarly to the default dependencies.
Currently, you can set Default Dependencies on 3 different layers:
- Component
- Family
- Project Configuration
This list also shows you the default override priority. If a component does not have default dependencies set, but its family has, the value from the value will be used.

I think we should take a similar approach with the platforms, where if a component does not have platform values set, but the project configuration does, when the packages are generated for the component, they should have the project configuration platforms values.

It would also be amazing if in the dropdown menu of the Component's platforms, we could have by default the value `project default` in addition to a `nil` value. The `nil` value basically tells the project to ignore the project configuration, but a `project default` would use it.
I hope I'm explaining this clearly.

What do you think?

File: Phoenix/PhoenixDocument+Extensions.swift (hunk 2 above)
Comment (serjooo) [reply Phoenix/PhoenixDocument+Extensions.swift]: Reverted in 41fd94518c094dc26f933268684b7019677128fd

File: Modules/Entities/PhoenixDocument/Sources/PhoenixDocument/Component/ProjectConfiguration.swift (hunk 1 above)
Comment (serjooo) [reply Modules/Entities/PhoenixDocument/Sources/PhoenixDocument/Component/ProjectConfiguration.swift]: Resolved in f0f74230e63634ac7f4cef39889488c45a0fbd6c

**Comment by serjooo** [comment]
>If I understood correctly from the code, your approach makes the change you make in the project configuration apply to all the components in the project. So when you update the iOS Version in Project Configuration, all components will have this new iOS Version set to them.
That's correct currently, we set the default to all of them. 

I understand the requirements, basically to achieve this we would need to the following:

- Add `PlatformsEditingView` to also in the `Family` Configuration
- Add `projectDefault` to platforms `iOSVersion`, `macOSVersion`, `tvOSVersion` etc..

I'm just not clear on the `nil` value. I'm guessing we want to use `nil` to ignore the `ProjectConfiguration` value. Maybe instead of `nil` we can have `ignoreProjectDefault` so it is a little bit more clear? 

Moreover, instead of adding the `projectDefault` or `ignoreProjectDefault`. I might suggested to make it into a checkbox as part of the platforms. Where you have three checkbox on the `Component`

[] Use Project Configuration Default
[] Use Family Configuration Default
[] Ignore Default Values

When tapping on `Ignore Default Value` we enable / show the `PlatformsEditingView`. That might result into a more straightforward UI/UX. What do you think?

Let me know if I understood the ideas correctly.

**Comment by Tawa** [comment]
No I don't think we need the have platforms defined in the family also. I just meant that the project configuration platforms should override the component platforms in case the component doesn't have values for them.

And as an extra, it would be amazing if we could allow the components to choose to ignore the project configuration even if they have a nil value.

**Comment by Tawa** [comment]
Maybe instead of making the values in the Platforms struct Optional, we could use something like the following:

```swift
fileprivate enum OptionalOverrideConstant {
    static let ignoreProjectValue: String = "ignoreProjectConfiguration"
}

public enum OptionalOverride<Value: RawRepresentable>: Codable where Value.RawValue: Codable {
    case projectDefault
    case ignoreProjectConfiguration
    case some(Value)
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if container.decodeNil() {
            self = .projectDefault
        } else if (try container.decode(String.self)) == OptionalOverrideConstant.ignoreProjectValue {
            self = .ignoreProjectConfiguration
        } else {
            let rawValue = try container.decode(Value.RawValue.self)
            let value = Value(rawValue: rawValue)
            self = value.flatMap(Self.some) ?? .projectDefault
        }
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        
        switch self {
        case .projectDefault:
            try container.encodeNil()
        case .ignoreProjectConfiguration:
            try container.encode(OptionalOverrideConstant.ignoreProjectValue)
        case .some(let value):
            try container.encode(value.rawValue)
        }
    }
}
```
And then we could disable the option to choose `ignoreProjectConfiguration` from the Project Confirugation View.

This is just an idea, you don't have to implement it in this PR. I think for not, simply making the component use the project configuration in case it doesn't have a value should be enough for your use case.

File: Modules/Generators/ProjectGenerator/Sources/ProjectGenerator/ComponentPackageProvider.swift [hunk 3]
```diff
@@ -170,11 +170,11 @@ public struct ComponentPackageProvider: ComponentPackageProviderProtocol {
         
         return .init(package: .init(name: packageName,
                                     defaultLocalization: defaultLocalization,
-                                    iOSVersion: component.iOSVersion,
-                                    macCatalystVersion: component.macCatalystVersion,
-                                    macOSVersion: component.macOSVersion,
-                                    tvOSVersion: component.tvOSVersion,
-                                    watchOSVersion: component.watchOSVersion,
+                                    iOSVersion: component.platforms.iOSVersion,
```
Comment (Tawa) [inline Modules/Generators/ProjectGenerator/Sources/ProjectGenerator/ComponentPackageProvider.swift:173]:
```suggestion
                                    iOSVersion: component.platforms.iOSVersion ?? projectConfiguration.platforms.iOSVersion,
```

This change for all those lines should be enough to make this work. You can revert injecting the document in the ProjectConfigurationView after doing this.

Comment (serjooo) [reply Modules/Generators/ProjectGenerator/Sources/ProjectGenerator/ComponentPackageProvider.swift]: Can you check [57ff8b8](https://github.com/Tawa/Phoenix/pull/74/commits/57ff8b8a9258fca074db8172f68ae99a2814d1c6) and [5f8e9b0](https://github.com/Tawa/Phoenix/pull/74/commits/5f8e9b08e4faa05dd5cf591617b71b45389e4dd5)?

**Review by Tawa — APPROVED** [review]

**Review by Tawa — APPROVED** [review]

**Pull request merged**